powervs: add podvm image documentation for IBM Cloud PowerVS provider

Amulyam24 opened the pull request:
This PR adds the instructions to build a pod VM image to be used with IBM Cloud PowerVS provider.

Fixes: #785

Amulyam24 commented:
The PR is ready for review. PTAL @yoheiueda @bpradipt

yoheiueda commented on podvm/Makefile.inc:
```diff
@@ -19,8 +19,9 @@ ARCH        ?= $(HOST_ARCH)
 DEB_ARCH    := $(subst x86_64,amd64,$(ARCH))
-LIBC        ?= $(if $(findstring s390x,$(ARCH)),gnu,musl)
-RUST_TARGET := $(ARCH)-unknown-linux-$(LIBC)
+LIBC        ?= $(if $(findstring s390x,$(ARCH)),gnu,$(if $(findstring ppc64le,$(ARCH)),gnu,musl))
```
You can simplify this like below.
```
LIBC        ?= $(if $(filter $(ARCH),s390x ppc64le),gnu,musl)
```

yoheiueda commented on podvm/Makefile.inc:
```diff
@@ -19,8 +19,9 @@ ARCH        ?= $(HOST_ARCH)
-LIBC        ?= $(if $(findstring s390x,$(ARCH)),gnu,musl)
-RUST_TARGET := $(ARCH)-unknown-linux-$(LIBC)
+LIBC        ?= $(if $(findstring s390x,$(ARCH)),gnu,$(if $(findstring ppc64le,$(ARCH)),gnu,musl))
+RUST_ARCH   ?= $(if $(findstring ppc64le,$(ARCH)),powerpc64le,$(ARCH))
```
ou can simplify this like below.
```
RUST_ARCH   ?= $(subst ppc64le,powerpc64le,$(ARCH))
```

yoheiueda commented on podvm/Makefile.inc:
```diff
@@ -116,8 +125,8 @@ $(UMOCI_SRC):
 
-$(PAUSE_SRC): $(SKOPEO_SRC)/bin/skopeo
-	$(SKOPEO_SRC)/bin/skopeo --override-arch $(DEB_ARCH) --policy "$(FILES_DIR)/etc/containers/policy.json" copy "$(PAUSE_REPO):$(PAUSE_VERSION)" "oci:$(PAUSE_SRC):$(PAUSE_VERSION)"
+$(PAUSE_SRC):
+	$(SKOPEO_BIN) --override-arch $(DEB_ARCH) --policy "$(FILES_DIR)/etc/containers/policy.json" copy "$(PAUSE_REPO):$(PAUSE_VERSION)" "oci:$(PAUSE_SRC):$(PAUSE_VERSION)"
```
```
$(PAUSE_SRC): $(SKOPEO_BIN)
	$(SKOPEO_BIN) --override-arch $(DEB_ARCH) --policy "$(FILES_DIR)/etc/containers/policy.json" copy "$(PAUSE_REPO):$(PAUSE_VERSION)" "oci:$(PAUSE_SRC):$(PAUSE_VERSION)"
```

yoheiueda commented on podvm/Makefile.inc:
```diff
@@ -80,6 +83,11 @@ PAUSE_VERSION = 3.6
 # Allows to override
 PAUSE_BUNDLE  ?= pause_bundle
 
+ifneq ($(ARCH), ppc64le)
+ 	BINARIES += $(SKOPEO)
+endif
+BINARIES += $(PAUSE)
+
```
We don't need this change.

yoheiueda commented on podvm/Makefile.inc:
```diff
@@ -106,8 +114,9 @@ $(STATIC_LIBSECCOMP):
 	git clone -b "v$(SKOPEO_VERSION)" "$(SKOPEO_REPO)" "$(SKOPEO_SRC)"
 
-$(SKOPEO_SRC)/bin/skopeo: $(SKOPEO_SRC)
+$(SKOPEO): $(SKOPEO_SRC)
 	cd "$(SKOPEO_SRC)" && CC= $(MAKE) bin/skopeo
+	$(eval SKOPEO_BIN := $(SKOPEO_SRC)/bin/skopeo)
```
We don't need this change.

yoheiueda commented on podvm/Makefile.inc:
```diff
@@ -66,8 +67,10 @@ AGENT_PROTOCOL_FORWARDER_SRC = ../..
 KATA_AGENT_BUILD_TYPE = release
 
-SKOPEO_SRC  = skopeo
+SKOPEO      = skopeo
+SKOPEO_SRC  = skopeo-src
 SKOPEO_REPO = https://github.com/containers/skopeo
+SKOPEO_BIN ?= $(shell command -v skopeo)
```
Could you change these line like this?

```
SKOPEO      = skopeo
SKOPEO_REPO = https://github.com/containers/skopeo
SKOPEO_BIN = $(SKOPEO)/bin/skopeo
```

yoheiueda commented on ibmcloud-powervs/image/Makefile:
```diff
@@ -0,0 +1,22 @@
+# SPDX-License-Identifier: Apache-2.0
+#
+include ../../podvm/Makefile.inc
+
```
You can reduce changes in Make.inc and .gitignore by overriding `SKOPEO_BIN` here like this.

```
SKOPEO_BIN = /usr/bin/skopeo
```

Amulyam24 replied:
Thanks for the suggestion @yoheiueda. I made the changes, PTAL!

bpradipt commented on ibmcloud-powervs/README.md:
````diff
@@ -0,0 +1,76 @@
+make build
+```
+
+> If you intend to use DHCP netowrk type for creating peer pod VMs with
````
nit: netowrk -> network

yoheiueda commented on podvm/Makefile.inc:
```diff
@@ -102,11 +104,9 @@ $(KATA_AGENT): force $(STATIC_LIBSECCOMP)
-# Skoepo package packages are available in RHEL/CentOS 8 or later and Ubuntu 20.10 or later
-$(SKOPEO_SRC):
+# Skopeo package packages are available in RHEL/CentOS 8 or later and Ubuntu 20.10 or later
+$(SKOPEO_BIN):
```
Clould you keep two rules separated, instead of merging them?  I think this is consistent to the rules for other binaries.

```
$(SKOPEO_SRC):
	git clone -b "v$(SKOPEO_VERSION)" "$(SKOPEO_REPO)" "$(SKOPEO_SRC)"

$(SKOPEO_BIN): $(SKOPEO_SRC)
	cd "$(SKOPEO_SRC)" && CC= $(MAKE) bin/skopeo
```

Amulyam24 replied:
Sure, Addressed the change.

yoheiueda commented on Makefile:
```diff
@@ -23,7 +23,7 @@ TEST_E2E_TIMEOUT ?= 20m
 	BUILTIN_CLOUD_PROVIDERS ?= aws azure ibmcloud vsphere
 else
-	BUILTIN_CLOUD_PROVIDERS ?= aws azure ibmcloud vsphere libvirt
+	BUILTIN_CLOUD_PROVIDERS ?= aws azure ibmcloud ibmcloud_powervs vsphere libvirt
```
@bpradipt Do we need some CI tests before adding this to the default provider list?

https://github.com/confidential-containers/cloud-api-adaptor/pull/750#discussion_r1153953076

bpradipt replied:
Hmm yeah at least some offline runs will be good. 

Tagging @stevenhorsman and @wainersm if they have anything to add.

stevenhorsman replied:
Yeah, I'll echo that. I don't think we should mandate publicly published test results for getting this PR merged (though we probably should before declaring any 'support statement'), but I think it would be great to get the new provider created in the e2e tests.

@Amulyam24 - there is some documentation on how to add a new provider in https://github.com/confidential-containers/cloud-api-adaptor/tree/staging/test/e2e#adding-support-for-a-new-cloud-provider, and there are examples of how to run tests in the `ibmcloud` provisioner code, but please reach out to us if you hit any issues as not many people have been through this process, so it might be difficult to follow and that's on us to try and improve!

Amulyam24 replied:
Thanks @stevenhorsman, I'll start looking into it and work on getting the test runs. Apart from that, is there anything pending to get this PR merged?

yoheiueda reviewed (approved):
LGTM. Thank you very much!

Amulyam24 commented:
Hi @stevenhorsman @yoheiueda @bpradipt,

After some initial analysis and inputs from [this discussion](https://cloud-native.slack.com/archives/C04A2EJ70BX/p1685694092890229), we have identified some dependencies and challenges in achieving a successful `ibmcloud-powervs` E2E flow.

1. Workload Deployments: The payload image `runtime-payload-ci`, built on top of the kata payload, is not being published for `ppc64le`. This includes support for kata-deploy.

2. Cluster Provisioning Challenge: We are currently exploring several options for creating a cluster on PowerVS. However, this might not be straightforward within the existing framework, and we are still deciding on the best approach.

3. Minor Challenge: Support for the `ppc64le` platform for the required images.

We are also aware of the ongoing effort regarding the [use of Docker for building VM images](https://github.com/confidential-containers/cloud-api-adaptor/issues/1006). If this is now the recommended approach for users, we can make the necessary changes for `ibmcloud-powervs` as well.

Taking the above points into consideration, could you please review and consider merging this PR? I can create the necessary issues in the applicable repositories and work on them.

Please share your thoughts and ideas on how we can handle this more effectively.

Thank you!

stevenhorsman commented:
> Taking the above points into consideration, could you please review and consider merging this PR? I can create the necessary issues in the applicable repositories and work on them.

Firstly - I'm sorry that I missed you message in slack, I've replied now, but in future feel free to ping me if you don't get a response.

I've commented in slack about the payloads and tried to answer your question and it looks like there are a bunch of pre-reqs that need resolving before you can get the operator install working, so raising issues for these makes sense, so please go ahead. However in light of this I think we need to make some changes to this PR before merging it. For example at the moment it adds `ibmcloud-powervs` as a default BUILTIN_CLOUD_PROVIDERS in the Makefile and to the documented list of supported cloud-providers in the README and I'm not sure how we can give those statements when install doesn't work. However I think it's just those changes that are problematic and if they are removed until the point that e2e tests can be run on power I don't see a big issue with merging the rest of the code (after the 0.6 release code freeze is lifted).

Does that help? Any other questions/comments on this?

Amulyam24 commented:
> Firstly - I'm sorry that I missed you message in slack, I've replied now, but in future feel free to ping me if you don't get a response.

No problem and sure, I will!

> I've commented in slack about the payloads and tried to answer your question and it looks like there are a bunch of pre-reqs that need resolving before you can get the operator install working, so raising issues for these makes sense, so please go ahead. However in light of this I think we need to make some changes to this PR before merging it. For example at the moment it adds `ibmcloud-powervs` as a default BUILTIN_CLOUD_PROVIDERS in the Makefile and to the documented list of supported cloud-providers in the README and I'm not sure how we can give those statements when install doesn't work. However I think it's just those changes that are problematic and if they are removed until the point that e2e tests can be run on power I don't see a big issue with merging the rest of the code (after the 0.6 release code freeze is lifted).
> 
> Does that help? Any other questions/comments on this?

Makes sense @stevenhorsman. I removed them and rebased the PR. PTAL!

stevenhorsman commented on ibmcloud-powervs/README.md:
````diff
@@ -0,0 +1,76 @@
+
+2. Download the qcow2 image and converts into ova type
+```
+pvsadm image qcow2ova --image-name <name> --image-dist centos --image-url https://cloud.centos.org/centos/8-stream/ppc64le/images/CentOS-Stream-GenericCloud-8-20220125.1.ppc64le.qcow2 --prep-template image-prep.template
````
When I ran this I got:
```
E0609 03:03:48.561007   87888 root.go:103] check failed: exec: "qemu-img": executable file not found in $PATH
Hint: yum install qemu-img -y
```
I assume that qemu-img (on fedora-like)/qemu-utils (for debian-like) needs to be added to the pre-reqs?

Amulyam24 replied:
That's right, I'll mention it under prereqs.

stevenhorsman commented on ibmcloud-powervs/README.md:
````diff
@@ -0,0 +1,76 @@
+
+2. Download the qcow2 image and converts into ova type
+```
+pvsadm image qcow2ova --image-name <name> --image-dist centos --image-url https://cloud.centos.org/centos/8-stream/ppc64le/images/CentOS-Stream-GenericCloud-8-20220125.1.ppc64le.qcow2 --prep-template image-prep.template
````
https://cloud.centos.org/centos/8-stream/ppc64le/images/CentOS-Stream-GenericCloud-8-20220125.1.ppc64le.qcow2 404s, so I assume this needs updating. Is there a more stable version that can be used, or switch to the https://cloud.centos.org/centos/8-stream/ppc64le/images/CentOS-Stream-GenericCloud-8-latest.ppc64le.qcow2 version?

Amulyam24 replied:
`https://cloud.centos.org/centos/8-stream/ppc64le/images/CentOS-Stream-GenericCloud-8-20220125.1.ppc64le.qcow2`, this is a pretty old one and looks like it isn't available anymore. Using the latest - `https://cloud.centos.org/centos/8-stream/ppc64le/images/CentOS-Stream-GenericCloud-8-latest.ppc64le.qcow2` should be fine.

stevenhorsman commented on ibmcloud-powervs/README.md:
````diff
@@ -0,0 +1,76 @@
+
+2. Download the qcow2 image and converts into ova type
+```
+pvsadm image qcow2ova --image-name <name> --image-dist centos --image-url https://cloud.centos.org/centos/8-stream/ppc64le/images/CentOS-Stream-GenericCloud-8-20220125.1.ppc64le.qcow2 --prep-template image-prep.template
````
I picked  https://cloud.centos.org/centos/8-stream/ppc64le/images/CentOS-Stream-GenericCloud-8-latest.ppc64le.qcow2 to try and get passed this issues, but the command failed with:
```
failed while preparing the image for centos distro, err: script /setup.sh failed with exitstatus: 1, stdout: , stderr: fork/exec /setup.sh: exec format error
```
suggesting it was doing something on the wrong architecture? Any ideas?

Amulyam24 replied:
IIRC, using `pvsadm` to convert the qcow2 to ova type works only on ppc64le(CentOS/RHEL).

stevenhorsman replied:
ok, should that go into the pre-req as well?

stevenhorsman reviewed (commented):
Hey @Amulyam24 - I tried to create the powervs image locally, but hit some issues, can you help me test this out please? Thanks!

Amulyam24 commented:
@stevenhorsman, a quick side-note/disclaimer: the current podvm image build succeeds but the kata-agent needs a fix as it fails to pull the pause image. I have the fixes ready but yet to validate it. I am planning to resume testing it and shall confirm on how it goes on Monday.

stevenhorsman reviewed (approved):
I wasn't able to get this working as I don't have access to a power machine to run it, but the issues that I picked up have been addressed, so I'm happy for this to go in without e2e test as it just effects powervs. Thanks!

bpradipt reviewed (approved):
/lgtm
Thanks @Amulyam24

stevenhorsman commented on ibmcloud-powervs/README.md:
````diff
@@ -0,0 +1,78 @@
+```
+yum install -y gcc gcc-c++ git make
+git clone -b CCv0 https://github.com/kata-containers/kata-containers.git
+git clone -b staging https://github.com/confidential-containers/cloud-api-adaptor.git
````
@Amulyam24 - FYI, my colleague @mattarnoatibm spotted that this is using a reference to the staging branch, whereas we have updated to `main` and I this wasn't merged when that update happened.

Amulyam24 replied:
Thanks @stevenhorsman and @mattarnoatibm. I'll submit a fix for it.
